File: advanced/javascript/pyramids.js
function pyramid(n) {
  const colCount = 2 * n - 1;
  const midpoint = Math.floor(colCount / 2);
  for (let row = 0; row < n; row++) {
    let level = '';
    for (let col = 0; col < colCount; col++) {
      let range = (midpoint - row <= col) &&  (midpoint + row >= col);
      if (range) {
        level += '#';
      } else {
        level += ' ';
      }
    }
    console.log(level)
  }
}

function pyramidRecursive(n, row = 0, level = '') {
  const colCount = n * 2 - 1;
  const midpoint = Math.floor(colCount / 2);
  const pyramidRange = (midpoint - row <= level.length) && (midpoint + row >= level.length);
  if (row === n) {
    return;
  }

  if (level.length === colCount) {
    console.log(level)
    return pyramidRecursive(n, row + 1)
  }

  if (pyramidRange) {
    level += '#'
  } else {
    level += ' '
  }
  return pyramidRecursive(n, row, level)
}

pyramidRecursive(9)

File: advanced/python/pyramids.py
from math import floor

def pyramid(n):
  col_count = n * 2 - 1
  midpoint = floor(col_count /2)
  for row in range(n):
    level = ''
    for col in range(col_count):
      pyramid_range = (midpoint - row <= col) and (midpoint + row >= col)
      if pyramid_range:
        level += '#'
      else:
        level += ' '
    print(level)

def pyramidRecursive(n, row = 0, level = ''):
  colCount = n * 2 - 1
  midpoint = floor(colCount / 2)
  level_length = len(level)
  pyramid_range = (midpoint - row <= level_length) and (midpoint + row >= level_length)

  if row == n: return

  if level_length == colCount:
    print(level)
    return pyramidRecursive(n, row + 1)

  if pyramid_range:
    level += '#'
  else: level += ' '

  return pyramidRecursive(n, row, level)

pyramidRecursive(9)